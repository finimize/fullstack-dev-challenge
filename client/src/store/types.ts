export const UPDATE_DETAILS = 'UPDATE_DETAILS'
export const VALIDATE_DETAILS = 'VALIDATE_DETAILS'
export const NEXT_PAGE = 'NEXT_PAGE'
export const PREVIOUS_PAGE = 'PREVIOUS_PAGE'
export const TOGGLE_MODE = 'TOGGLE_MODE'
