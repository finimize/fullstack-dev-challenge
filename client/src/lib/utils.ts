import type { ChartTooltipItem } from 'chart.js'
import { SAVINGS_DEFAULTS } from './constants'

export const createQueryString = (queries: { [key: string]: any }): string =>
    Object.keys(queries)
        .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(queries[key]))
        .join('&')

export const appendYearsQuery = (queryString: string) =>
    `${queryString}&years=${SAVINGS_DEFAULTS.YEARS}`

export const extractYearlyData = (projections: number[]) =>
    projections.filter((item, index) => index % 12 === 0)

export const numberFormatter = new Intl.NumberFormat('gb-GB', {
    style: 'currency',
    currency: 'GBP',
})

export const getTooltipTitle = (tooltipItems: ChartTooltipItem[]) => `Year ${tooltipItems[0].label}`

export const getTooltipLabel = (tooltipItem: ChartTooltipItem) => {
    const value = tooltipItem.value ? parseFloat(tooltipItem.value) : 0
    return numberFormatter.format(value)
}
