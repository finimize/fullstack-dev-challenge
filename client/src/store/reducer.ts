import {
    UPDATE_DETAILS,
    UPDATE_CALCULATIONS,
    UPDATE_INTEREST,
    UPDATE_DATA,
    VALIDATE_DETAILS,
    NEXT_PAGE,
    TOGGLE_MODE,
    PREVIOUS_PAGE,
} from './types'
import {
    StateInterface,
    UpdateDetailsTypeInterface,
    UpdateCalculationsTypeInterface,
    UpdateInterestRateInterface,
    UpdateDataInterface,
    ValidateDetailsTypeInterface,
    NextPageTypeInterface,
    ToggleModeInterface,
    PreviousPageTypeInterface,
} from './store.interface'

export const reducer = (
    initialState: StateInterface,
    action:
        | UpdateDetailsTypeInterface
        | UpdateCalculationsTypeInterface
        | UpdateInterestRateInterface
        | UpdateDataInterface
        | ValidateDetailsTypeInterface
        | NextPageTypeInterface
        | PreviousPageTypeInterface
        | ToggleModeInterface,
): StateInterface => {
    switch (action.type) {
        case NEXT_PAGE: {
            return {
                ...initialState,
                currentPage: initialState.currentPage + 1,
            }
        }
        case PREVIOUS_PAGE: {
            return {
                ...initialState,
                currentPage: initialState.currentPage - 1,
            }
        }
        case TOGGLE_MODE: {
            return {
                ...initialState,
                calculatorMode: action.payload,
            }
        }
        case UPDATE_DETAILS:
            return {
                ...initialState,
                details: {
                    ...initialState.details,
                    [action.field]: action.payload,
                },
            }
        case UPDATE_CALCULATIONS:
            return {
                ...initialState,
                calculations: {
                    ...initialState.calculations,
                    [action.field]: action.payload,
                },
            }
        case UPDATE_INTEREST:
            return {
                ...initialState,
                calculations: {
                    ...initialState.calculations,
                    interestRate: action.payload,
                },
            }
        case UPDATE_DATA:
            return {
                ...initialState,
                data: {
                    ...initialState.data,
                    value: action.payload,
                },
            }
        case VALIDATE_DETAILS:
            return {
                ...initialState,
                details: {
                    ...initialState.details,
                    firstName: {
                        ...initialState.details.firstName,
                        error: action.payload.firstName,
                    },
                    lastName: {
                        ...initialState.details.lastName,
                        error: action.payload.lastName,
                    },
                    email: {
                        ...initialState.details.email,
                        error: action.payload.email,
                    },
                },
            }
        default:
            return initialState
    }
}
