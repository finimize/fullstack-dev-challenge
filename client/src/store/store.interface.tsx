import {
    UPDATE_DETAILS,
    UPDATE_CALCULATIONS,
    UPDATE_INTEREST,
    UPDATE_DATA,
    VALIDATE_DETAILS,
    NEXT_PAGE,
    TOGGLE_MODE,
    TOGGLE_COMPOUNDING_FREQUENCY,
    PREVIOUS_PAGE,
} from './types'

export interface TextInputInterface {
    value: string
    error: string
}

interface DetailsInterface {
    firstName: TextInputInterface
    lastName: TextInputInterface
    email: TextInputInterface
}

export interface DataValueInterface {
    yearlySavings: number[]
    finalValue: number
    yearlyBreakdown: {
        yearFinal: number
        savings: []
    }[]
}

export interface DataInterface {
    value: DataValueInterface | null
    error: string
    isLoading: boolean
}

export interface CalculationsInterface {
    initialSavings: TextInputInterface
    interestRate: number
    compoundingFrequency: number
    monthlyContributions: TextInputInterface
}

export interface StateInterface {
    details: DetailsInterface
    data: DataInterface
    currentPage: number
    calculatorMode: string
    calculations: CalculationsInterface
}

export type UpdateDetailsTypeInterface = {
    type: typeof UPDATE_DETAILS
    field: 'firstName' | 'lastName' | 'email'
    payload: Partial<TextInputInterface>
}
export type UpdateCalculationsTypeInterface = {
    type: typeof UPDATE_CALCULATIONS
    field: 'initialSavings' | 'monthlyContributions'
    payload: Partial<TextInputInterface>
}

export type UpdateInterestRateInterface = {
    type: typeof UPDATE_INTEREST
    payload: number
}

export type AppProps = {
    children: React.ReactNode
}

export type ValidateDetailsTypeInterface = {
    type: typeof VALIDATE_DETAILS
    payload: {
        firstName: string
        lastName: string
        email: string
    }
}

export type NextPageTypeInterface = {
    type: typeof NEXT_PAGE
}

export type PreviousPageTypeInterface = {
    type: typeof PREVIOUS_PAGE
}

export type ToggleModeInterface = {
    type: typeof TOGGLE_MODE
    payload: string
}

export enum CompoundingFrequency {
    annually = 1,
    quarterly = 4,
    monthly = 12,
}

export type ToggleCompoundingFrequencyInterface = {
    type: typeof TOGGLE_COMPOUNDING_FREQUENCY
    payload: string
}

export type UpdateDataInterface = {
    type: typeof UPDATE_DATA
    payload: DataInterface
}
