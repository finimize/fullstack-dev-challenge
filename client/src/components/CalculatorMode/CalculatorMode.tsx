import React, { FC, useContext } from 'react'
import { Box, Text, useRadio, UseRadioProps, useRadioGroup } from '@chakra-ui/react'
import { AppContext } from '../../store'
import { TOGGLE_MODE, NEXT_PAGE, PREVIOUS_PAGE } from '../../store/types'
import { Card } from '../Card'
import { PageControls } from '../PageControls'

interface RadioCardProps extends UseRadioProps {
    children: React.ReactNode
}

const RadioCard = (props: RadioCardProps) => {
    const { getInputProps, getCheckboxProps } = useRadio(props)

    const input = getInputProps()
    const checkbox = getCheckboxProps()

    return (
        <Box display='flex' as='label'>
            <input {...input} />
            <Box
                {...checkbox}
                cursor='pointer'
                borderWidth='1px'
                borderRadius='md'
                boxShadow='md'
                bg='white'
                _checked={{
                    bg: 'blue600',
                    color: 'white',
                    borderColor: 'white',
                    borderWidth: '2px',
                }}
                _focus={{
                    boxShadow: 'outline',
                }}
                px={5}
                py={3}
                width='150px'
                textAlign='center'
                fontSize='sm'
                fontWeight='700'
            >
                {props.children}
            </Box>
        </Box>
    )
}

export const CalculatorMode: FC = () => {
    const { state, dispatch } = useContext(AppContext)
    const options = ['Simple', 'Advanced']

    const { getRootProps, getRadioProps } = useRadioGroup({
        name: 'framework',
        defaultValue: state.calculatorMode,
        onChange: (value) => dispatch({ type: TOGGLE_MODE, payload: String(value) }),
    })

    const group = getRootProps()

    const nextPage = () => dispatch({ type: NEXT_PAGE })
    const prevPage = () => dispatch({ type: PREVIOUS_PAGE })
    return (
        <Box>
            <Card>
                <Text
                    fontSize='sm'
                    marginBottom='4'
                    color='blueHeader'
                    fontWeight='700'
                    textAlign='center'
                >
                    Choose how you want to see your savings
                </Text>
                <Text fontSize='xs' marginBottom='4' color='grey5' textAlign='center'>
                    Hi <em>{state.details.firstName.value}</em> Choose one of the options below to
                    to see how much money you can save:
                </Text>
                <Box backgroundColor='blue200' padding='8' margin='0 100px' borderRadius='8px'>
                    <Box {...group} display='flex' justifyContent='space-evenly' width='100%'>
                        {options.map((value) => {
                            const radio = getRadioProps({ value })
                            return (
                                <RadioCard key={value} {...radio}>
                                    {value}
                                </RadioCard>
                            )
                        })}
                    </Box>
                    <Box padding='2'>
                        <Text fontSize='sm' margin='4' color='blue800'>
                            {state.calculatorMode === 'Simple'
                                ? 'This will allow you to see excatly how much your money will grow over time as soon as you make your first deposit.'
                                : 'This advanced setting will let you picture how much you could posssibly make if you were contributing regularly to your account'}
                        </Text>
                    </Box>
                </Box>
            </Card>
            <PageControls prevPage={prevPage} nextPage={nextPage} />
        </Box>
    )
}
