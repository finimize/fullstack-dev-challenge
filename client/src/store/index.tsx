import React, { createContext, useReducer, FC } from 'react'
import { initialState } from './initialState'
import { reducer } from './reducer'
import {
    UpdateDetailsTypeInterface,
    ValidateDetailsTypeInterface,
    NextPageTypeInterface,
    StateInterface,
    AppProps,
    PreviousPageTypeInterface,
    ToggleModeInterface,
    ToggleCompoundingFrequencyInterface,
    UpdateCalculationsTypeInterface,
    UpdateInterestRateInterface,
    UpdateDataInterface,
} from './store.interface'

export type DispatchType =
    | UpdateDetailsTypeInterface
    | UpdateCalculationsTypeInterface
    | UpdateDataInterface
    | ValidateDetailsTypeInterface
    | NextPageTypeInterface
    | ToggleModeInterface
    | ToggleCompoundingFrequencyInterface
    | PreviousPageTypeInterface
    | UpdateInterestRateInterface
const AppContext = createContext<{
    state: StateInterface
    dispatch: React.Dispatch<DispatchType>
}>({
    state: initialState,
    dispatch: () => null,
})

const AppProvider: FC<AppProps> = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState)
    return <AppContext.Provider value={{ state, dispatch }}>{children}</AppContext.Provider>
}

export { AppContext, AppProvider }
