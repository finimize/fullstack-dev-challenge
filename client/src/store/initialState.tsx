import { StateInterface, CompoundingFrequency } from './store.interface'

export const initialState: StateInterface = {
    details: {
        firstName: {
            value: '',
            error: '',
        },
        lastName: {
            value: '',
            error: '',
        },
        email: {
            value: '',
            error: '',
        },
    },
    data: {
        value: null,
        error: '',
        isLoading: false,
    },
    calculations: {
        initialSavings: {
            value: '10',
            error: '',
        },
        interestRate: 20,
        compoundingFrequency: CompoundingFrequency.annually,
        monthlyContributions: {
            value: '0',
            error: '',
        },
    },
    calculatorMode: 'Simple',
    currentPage: 1,
}
