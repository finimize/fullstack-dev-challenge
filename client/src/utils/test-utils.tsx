import React, { FC, useReducer } from 'react'
import { ChakraProvider, extendTheme } from '@chakra-ui/react'
import { render as rtlRender, RenderResult, RenderOptions } from '@testing-library/react'
import { AppContext, DispatchType } from '../store'
import { initialState } from '../store/initialState'
import { reducer } from '../store/reducer'
import { StateInterface } from '../store/store.interface'
import { theme } from '../theme'

const mainTheme = extendTheme(theme)

interface CustomRenderOptions extends Partial<RenderOptions> {
    store?: StateInterface
    defaultTheme?: typeof theme
}

export interface CustomRenderInterface {
    (
        ui: React.ReactElement<unknown, string | React.JSXElementConstructor<unknown>>,
        options: CustomRenderOptions,
    ): { render: RenderResult; dispatchMock: jest.Mock<unknown, unknown[]> }
}

const customRender: CustomRenderInterface = (
    ui,
    { store = initialState, defaultTheme = mainTheme, ...options },
) => {
    const dispatchMock = jest.fn()
    const AllProviders: FC = ({ children }) => {
        // const [calculatorState, setCalculatorState] = useState(store)
        const [calculatorState, dispatch] = useReducer(reducer, store)
        return (
            <ChakraProvider theme={defaultTheme}>
                <AppContext.Provider
                    value={{
                        state: calculatorState,
                        dispatch: dispatchMock.mockImplementation((value: DispatchType) =>
                            dispatch(value),
                        ),
                    }}
                >
                    {' '}
                    {children}{' '}
                </AppContext.Provider>{' '}
            </ChakraProvider>
        )
    }
    return { render: rtlRender(ui, { wrapper: AllProviders, ...options }), dispatchMock }
}

export { customRender }
