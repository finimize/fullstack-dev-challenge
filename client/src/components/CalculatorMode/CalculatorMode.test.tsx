import React from 'react'
import { cleanup, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { initialState } from '../../store/initialState'
import { customRender } from '../../utils/test-utils'
import { CalculatorMode } from './CalculatorMode'

afterEach(cleanup)

describe('component/CalculatorMode', () => {
    it('renders Calculator Mode component', () => {
        const { render } = customRender(<CalculatorMode />, {})
        const { container } = render
        expect(container).toMatchSnapshot()
    })

    it('allows calculator modes to be changed to advacned mode', () => {
        const { render, dispatchMock } = customRender(<CalculatorMode />, {})
        const { getByText, getByTestId, container } = render
        expect(getByText('Simple')).toBeInTheDocument()
        const advancedbutton = getByTestId('Advanced')
        const simplebutton = getByTestId('Simple')
        // Check initial view
        expect(simplebutton).toHaveAttribute('checked')
        expect(advancedbutton).not.toBeChecked()
        // Check Advanced button is working
        fireEvent.click(advancedbutton)
        expect(dispatchMock).toHaveBeenCalledWith({ payload: 'Advanced', type: 'TOGGLE_MODE' })
        expect(advancedbutton).toBeChecked()
        expect(container).toMatchSnapshot()
    })

    it('allows calculator modes to be changed to simple mode', () => {
        const { render, dispatchMock } = customRender(<CalculatorMode />, {
            store: { ...initialState, calculatorMode: 'Advanced' },
        })
        const { getByText, getByTestId, container } = render
        expect(getByText('Simple')).toBeInTheDocument()
        const advancedbutton = getByTestId('Advanced')
        const simplebutton = getByTestId('Simple')
        // Check initial view
        expect(simplebutton).not.toHaveAttribute('checked')
        expect(advancedbutton).toBeChecked()
        // Check Advanced button is working
        fireEvent.click(simplebutton)
        expect(dispatchMock).toHaveBeenNthCalledWith(1, { payload: 'Simple', type: 'TOGGLE_MODE' })
        expect(dispatchMock).toHaveBeenNthCalledWith(2, {
            type: 'UPDATE_CALCULATIONS',
            field: 'monthlyContributions',
            payload: {
                value: '0',
            },
        })
        expect(dispatchMock).toHaveBeenNthCalledWith(3, {
            payload: 'annually',
            type: 'TOGGLE_COMPOUNDING_FREQUENCY',
        })
        expect(dispatchMock).toHaveBeenCalledTimes(3)
        expect(simplebutton).toBeChecked()
        expect(container).toMatchSnapshot()
    })
    it('allows to hit', () => {
        const { render, dispatchMock } = customRender(<CalculatorMode />, {
            store: { ...initialState, currentPage: 2 },
        })
        const { getByTestId } = render
        const prevbutton = getByTestId('previous-button')
        const nextbutton = getByTestId('next-button')
        userEvent.click(prevbutton)
        expect(dispatchMock).toHaveBeenCalledWith({ type: 'PREVIOUS_PAGE' })
        userEvent.click(nextbutton)
        expect(dispatchMock).toHaveBeenCalledWith({ type: 'NEXT_PAGE' })
    })
})
