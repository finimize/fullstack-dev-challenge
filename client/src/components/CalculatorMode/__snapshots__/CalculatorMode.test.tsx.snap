// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/CalculatorMode allows calculator modes to be changed to advacned mode 1`] = `
<div>
   
  <div
    class="css-0"
  >
    <div
      class="css-6z87o"
    >
      <p
        class="chakra-text css-nrvl4t"
      >
        Choose how you want to see your savings
      </p>
      <p
        class="chakra-text css-ulf43m"
      >
        Hi 
        <em />
         Choose one of the options below to to see how much money you can save:
      </p>
      <div
        class="css-1wlqeol"
      >
        <div
          class="css-ocze9x"
          role="radiogroup"
        >
          <label
            class="css-k008qs"
          >
            <input
              checked=""
              data-testid="Simple"
              name="framework"
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="radio"
              value="Simple"
            />
            <div
              aria-hidden="true"
              class="css-1r0ytuv"
            >
              Simple
            </div>
          </label>
          <label
            class="css-k008qs"
          >
            <input
              data-testid="Advanced"
              name="framework"
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="radio"
              value="Advanced"
            />
            <div
              aria-hidden="true"
              class="css-1r0ytuv"
              data-checked=""
            >
              Advanced
            </div>
          </label>
        </div>
        <div
          class="css-olyig7"
        >
          <p
            class="chakra-text css-8om1m1"
          >
            This advanced setting will let you picture how much you could posssibly make if you were contributing regularly to your account
          </p>
        </div>
      </div>
    </div>
    <div
      class="css-j4xgao"
    >
      <button
        class="chakra-button css-1mi4eb0"
        data-testid="next-button"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
   
   
</div>
`;

exports[`component/CalculatorMode allows calculator modes to be changed to simple mode 1`] = `
<div>
   
  <div
    class="css-0"
  >
    <div
      class="css-6z87o"
    >
      <p
        class="chakra-text css-nrvl4t"
      >
        Choose how you want to see your savings
      </p>
      <p
        class="chakra-text css-ulf43m"
      >
        Hi 
        <em />
         Choose one of the options below to to see how much money you can save:
      </p>
      <div
        class="css-1wlqeol"
      >
        <div
          class="css-ocze9x"
          role="radiogroup"
        >
          <label
            class="css-k008qs"
          >
            <input
              data-testid="Simple"
              name="framework"
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="radio"
              value="Simple"
            />
            <div
              aria-hidden="true"
              class="css-1r0ytuv"
              data-checked=""
            >
              Simple
            </div>
          </label>
          <label
            class="css-k008qs"
          >
            <input
              checked=""
              data-testid="Advanced"
              name="framework"
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="radio"
              value="Advanced"
            />
            <div
              aria-hidden="true"
              class="css-1r0ytuv"
            >
              Advanced
            </div>
          </label>
        </div>
        <div
          class="css-olyig7"
        >
          <p
            class="chakra-text css-8om1m1"
          >
            This will allow you to see excatly how much your money will grow over time as soon as you make your first deposit.
          </p>
        </div>
      </div>
    </div>
    <div
      class="css-j4xgao"
    >
      <button
        class="chakra-button css-1mi4eb0"
        data-testid="next-button"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
   
   
</div>
`;

exports[`component/CalculatorMode renders Calculator Mode component 1`] = `
<div>
   
  <div
    class="css-0"
  >
    <div
      class="css-6z87o"
    >
      <p
        class="chakra-text css-nrvl4t"
      >
        Choose how you want to see your savings
      </p>
      <p
        class="chakra-text css-ulf43m"
      >
        Hi 
        <em />
         Choose one of the options below to to see how much money you can save:
      </p>
      <div
        class="css-1wlqeol"
      >
        <div
          class="css-ocze9x"
          role="radiogroup"
        >
          <label
            class="css-k008qs"
          >
            <input
              checked=""
              data-testid="Simple"
              name="framework"
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="radio"
              value="Simple"
            />
            <div
              aria-hidden="true"
              class="css-1r0ytuv"
              data-checked=""
            >
              Simple
            </div>
          </label>
          <label
            class="css-k008qs"
          >
            <input
              data-testid="Advanced"
              name="framework"
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="radio"
              value="Advanced"
            />
            <div
              aria-hidden="true"
              class="css-1r0ytuv"
            >
              Advanced
            </div>
          </label>
        </div>
        <div
          class="css-olyig7"
        >
          <p
            class="chakra-text css-8om1m1"
          >
            This will allow you to see excatly how much your money will grow over time as soon as you make your first deposit.
          </p>
        </div>
      </div>
    </div>
    <div
      class="css-j4xgao"
    >
      <button
        class="chakra-button css-1mi4eb0"
        data-testid="next-button"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
   
   
</div>
`;
